samples/snippets/src/main/java/com/example/bigquerystorage/WriteToDefaultStream.java

```diff
@@ -45,36 +51,119 @@ public static void runWriteToDefaultStream()
+      Table table = bigquery.getTable(parentTable.getDataset(), parentTable.getTable());
+      Schema schema = table.getDefinition().getSchema();
+      TableSchema tableSchema = BqToBqStorageSchemaConverter.convertTableSchema(schema);
+
+      // Use the JSON stream writer to send records in JSON format. Specify the table name to write
+      // to the default stream.
+      // For more information about JsonStreamWriter, see:
+      // https://googleapis.dev/java/google-cloud-bigquerystorage/latest/com/google/cloud/bigquery/storage/v1/JsonStreamWriter.html
+      streamWriter = JsonStreamWriter.newBuilder(parentTable.toString(), tableSchema).build();
+    }
+
+    public void append(JSONArray data) throws DescriptorValidationException, IOException {
+      synchronized (this.lock) {
+        // If earlier appends have failed, we need to reset before continuing.
+        if (this.error != null) {
+          throw this.error;
+        }
+      }
+      // Append asynchronously for increased throughput.
+      ApiFuture<AppendRowsResponse> future = streamWriter.append(data);
+      ApiFutures.addCallback(
+          future, new AppendCompleteCallback(this), MoreExecutors.directExecutor());
+      // Increase the count of in-flight requests.
+      inflightRequestCount.register();
+    }
+
+    public void cleanup() {
+      // Wait for all in-flight requests to complete.
+      inflightRequestCount.arriveAndAwaitAdvance();
+
+      // Close the connection to the server.
+      streamWriter.close();
+
+      // Verify that no error occurred in the stream.
+      synchronized (this.lock) {
+        if (this.error != null) {
+          throw this.error;
+        }
+      }
+    }
+
+    static class AppendCompleteCallback implements ApiFutureCallback<AppendRowsResponse> {
+
+      private final DataWriter parent;
+
+      public AppendCompleteCallback(DataWriter parent) {
+        this.parent = parent;
+      }
+
+      public void onSuccess(AppendRowsResponse response) {
+        System.out.format("Append success\n");
+        done();
+      }
+
+      public void onFailure(Throwable throwable) {
+        // If the wrapped exception is a StatusRuntimeException, check the state of the operation.
+        // If the state is INTERNAL, CANCELLED, or ABORTED, you can retry. For more information,
+        // see: https://grpc.github.io/grpc-java/javadoc/io/grpc/StatusRuntimeException.html
+        synchronized (this.parent.lock) {
+          if (this.parent.error == null) {
+            StorageException storageException = Exceptions.toStorageException(throwable);
+            this.parent.error =
+                (storageException != null) ? storageException : new RuntimeException(throwable);
+          }
         }
-        ApiFuture<AppendRowsResponse> future = writer.append(jsonArr);
-        AppendRowsResponse response = future.get();
+        System.out.format("Error: %s\n", throwable.toString());
```

Comment:
Could we add some simple immediate retry pattern here?

Reply:
Done